File: scripts/validate.py
#!/usr/bin/env python3
"""Validate cartitas card files for basic kid-content rules."""

from __future__ import annotations

import re
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent.parent

CARD_TYPES = re.compile(r"^(Q:|A:|C:)", re.MULTILINE)
MAX_ANSWER_LEN = 500  # soft limit for kid answers


def validate_file(path: Path) -> list[str]:
    errors: list[str] = []
    text = path.read_text(encoding="utf-8")
    rel = path.relative_to(ROOT)

    if not text.strip():
        errors.append(f"{rel}: file is empty")
        return errors

    blocks = [b.strip() for b in re.split(r"^---\s*$", text, flags=re.MULTILINE) if b.strip()]
    for i, block in enumerate(blocks, start=1):
        has_q = any(line.startswith("Q:") for line in block.splitlines())
        has_a = any(line.startswith("A:") for line in block.splitlines())
        has_c = any(line.startswith("C:") for line in block.splitlines())

        if has_c and (has_q or has_a):
            errors.append(f"{rel} card {i}: cloze (C:) mixed with Q/A in same block")
        elif has_q != has_a:
            errors.append(f"{rel} card {i}: Q/A pair incomplete")
        elif not has_c and not has_q:
            # Allow section-only blocks at file start (headers before first card)
            if i == 1 and block.startswith("#"):
                continue
            errors.append(f"{rel} card {i}: no Q/A or C: card found")

        for line in block.splitlines():
            if line.startswith("A:") and len(line) > MAX_ANSWER_LEN:
                errors.append(f"{rel} card {i}: answer very long ({len(line)} chars)")

    return errors


def main() -> int:
    dirs = [ROOT / "cards", ROOT / "queue", ROOT / "draft"]
    all_errors: list[str] = []

    for base in dirs:
        if not base.is_dir():
            continue
        for path in sorted(base.rglob("*.md")):
            all_errors.extend(validate_file(path))

    if all_errors:
        for err in all_errors:
            print(err, file=sys.stderr)
        return 1

    print("All card files passed basic validation.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: internal/validate/validate.go
package validate

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/qu1queee/cartitas/internal/cardsplit"
)

const maxAnswerLen = 500

func File(root, path string) []string {
	var errors []string
	data, err := os.ReadFile(path)
	if err != nil {
		return []string{fmt.Sprintf("%s: %v", rel(root, path), err)}
	}
	text := string(data)
	relPath := rel(root, path)

	if strings.TrimSpace(text) == "" {
		return []string{fmt.Sprintf("%s: file is empty", relPath)}
	}

	blocks := cardsplit.Split(text)
	for i, block := range blocks {
		n := i + 1
		hasQ, hasA, hasC := false, false, false
		for _, line := range strings.Split(block, "\n") {
			switch {
			case strings.HasPrefix(line, "Q:"):
				hasQ = true
			case strings.HasPrefix(line, "A:"):
				hasA = true
				if len(line) > maxAnswerLen {
					errors = append(errors, fmt.Sprintf("%s card %d: answer very long (%d chars)", relPath, n, len(line)))
				}
			case strings.HasPrefix(line, "C:"):
				hasC = true
			}
		}

		if hasC && (hasQ || hasA) {
			errors = append(errors, fmt.Sprintf("%s card %d: cloze (C:) mixed with Q/A in same block", relPath, n))
		} else if hasQ != hasA {
			errors = append(errors, fmt.Sprintf("%s card %d: Q/A pair incomplete", relPath, n))
		} else if !hasC && !hasQ {
			if n == 1 && strings.HasPrefix(block, "#") {
				continue
			}
			errors = append(errors, fmt.Sprintf("%s card %d: no Q/A or C: card found", relPath, n))
		}
	}
	return errors
}

func All(root string) []string {
	var all []string
	for _, dir := range []string{"cards", "queue", "draft"} {
		base := filepath.Join(root, dir)
		info, err := os.Stat(base)
		if err != nil || !info.IsDir() {
			continue
		}
		var files []string
		_ = filepath.WalkDir(base, func(path string, d os.DirEntry, err error) error {
			if err != nil || d.IsDir() || !strings.HasSuffix(path, ".md") {
				return nil
			}
			files = append(files, path)
			return nil
		})
		sort.Strings(files)
		for _, path := range files {
			all = append(all, File(root, path)...)
		}
	}
	return all
}

func rel(root, path string) string {
	r, err := filepath.Rel(root, path)
	if err != nil {
		return path
	}
	return filepath.ToSlash(r)
}
